File: electron/projects.ts
import * as crypto from "crypto";
import * as fs from "fs";
import * as path from "path";
import type { z } from "zod";
import type { Database } from "better-sqlite3";
import { ProjectConfigSchema } from "../src/types/schemas";
import type { Project, ProjectConfig, Provider } from "../src/types/index";

// ─── Helpers ─────────────────────────────────────────────────────────────────

function configPath(projectPath: string): string {
  return path.join(projectPath, ".aichemist", "config.json");
}

function defaultProjectConfig(defaultProvider: Provider = "anthropic"): ProjectConfig {
  return {
    provider: defaultProvider,
    model: defaultProvider === "anthropic" ? "claude-sonnet-4-6" : "",
    approval_mode: "custom",
    approval_rules: [
      { tool_category: "filesystem", policy: "risky_only" },
      { tool_category: "shell", policy: "always" },
      { tool_category: "web", policy: "never" },
    ],
    custom_tools: [],
    allowed_tools: [],
    create_worktree_per_session: false,
  };
}

function valueAtPath(value: unknown, issuePath: PropertyKey[]): unknown {
  return issuePath.reduce<unknown>((current, key) => {
    if (current === null || current === undefined) return undefined;
    if (typeof current !== "object") return undefined;
    return (current as Record<PropertyKey, unknown>)[key];
  }, value);
}

function summarizeConfigValue(value: unknown): unknown {
  if (value === null || value === undefined) return value;
  if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") return value;
  if (Array.isArray(value)) return `[array:${value.length}]`;
  if (typeof value === "object") return `[object:${Object.keys(value).join(",")}]`;
  return typeof value;
}

function describeProjectConfigIssue(issue: z.ZodIssue, parsed: unknown): Record<string, unknown> {
  const details: Record<string, unknown> = {
    path: issue.path.length > 0 ? issue.path.join(".") : "<root>",
    code: issue.code,
    message: issue.message,
  };
  const actual = valueAtPath(parsed, issue.path);
  if (actual !== undefined) {
    details.actual = summarizeConfigValue(actual);
  }
  if ("values" in issue) {
    details.expected = issue.values;
  } else if ("options" in issue) {
    details.expected = (issue as { options: unknown }).options;
  }
  return details;
}

/**
 * Parse raw JSON into a validated ProjectConfig.
 * Unknown extra fields are stripped; missing optional fields get defaults.
 * Returns the default config if validation fails (corrupt/incompatible file).
 */
function parseProjectConfig(raw: string, sourcePath?: string): ProjectConfig {
  try {
    const parsed = JSON.parse(raw);
    const result = ProjectConfigSchema.safeParse(parsed);
    if (result.success) {
      const config = result.data;
      if (!config.model && config.provider === "anthropic") {
        config.model = "claude-sonnet-4-6";
      }
      return config;
    }
    const warning = {
      configPath: sourcePath,
      issues: result.error.issues.map((issue) => describeProjectConfigIssue(issue, parsed)),
    };
    console.warn(
      "[projects] ProjectConfig validation failed, falling back to defaults:",
      JSON.stringify(warning, null, 2)
    );
    return defaultProjectConfig();
  } catch {
    return defaultProjectConfig();
  }
}

function readOrCreateConfig(projectPath: string, defaultProvider: Provider = "anthropic"): ProjectConfig {
  const cfgPath = configPath(projectPath);
  if (fs.existsSync(cfgPath)) {
    try {
      const raw = fs.readFileSync(cfgPath, "utf-8");
      return parseProjectConfig(raw, cfgPath);
    } catch {
      return defaultProjectConfig();
    }
  }

  const config = defaultProjectConfig(defaultProvider);
  const dir = path.dirname(cfgPath);
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(cfgPath, JSON.stringify(config, null, 2), "utf-8");
  return config;
}

function nowIso(): string {
  return new Date().toISOString();
}

// ─── Exported functions ──────────────────────────────────────────────────────

/**
 * Open a folder as a project. Creates `.aichemist/config.json` if absent.
 */
export function addProject(db: Database, projectPath: string, defaultProvider: Provider = "anthropic"): Project {
  const trimmedPath = projectPath.replace(/\/+$/, "");

  if (!fs.existsSync(trimmedPath) || !fs.statSync(trimmedPath).isDirectory()) {
    throw new Error(`Path is not a directory: ${trimmedPath}`);
  }

  const name = path.basename(trimmedPath) || "Project";
  const config = readOrCreateConfig(trimmedPath, defaultProvider);
  const id = crypto.randomUUID();
  const createdAt = nowIso();

  // INSERT OR IGNORE so re-opening an existing project is a no-op on the DB level
  db.prepare(
    "INSERT OR IGNORE INTO projects (id, name, path, created_at) VALUES (?, ?, ?, ?)"
  ).run(id, name, trimmedPath, createdAt);

  // Fetch the canonical row (handles the case where the project already existed)
  const row = db
    .prepare("SELECT id, name, path, created_at FROM projects WHERE path = ?")
    .get(trimmedPath) as { id: string; name: string; path: string; created_at: string };

  return {
    id: row.id,
    name: row.name,
    path: row.path,
    created_at: row.created_at,
    config,
  };
}

/**
 * Return all known projects, ordered by creation date.
 */
export function listProjects(db: Database): Project[] {
  const rows = db
    .prepare("SELECT id, name, path, created_at FROM projects ORDER BY created_at ASC")
    .all() as { id: string; name: string; path: string; created_at: string }[];

  return rows.map((row) => {
    const config = readOrCreateConfig(row.path);
    return {
      id: row.id,
      name: row.name,
      path: row.path,
      created_at: row.created_at,
      config,
    };
  });
}

/**
 * Remove a project from the registry. Does NOT delete the folder.
 */
export function removeProject(db: Database, id: string): void {
  db.prepare("DELETE FROM projects WHERE id = ?").run(id);
}

/**
 * Read `.aichemist/config.json` for a project.
 */
export function getProjectConfig(db: Database, id: string): ProjectConfig {
  const row = db
    .prepare("SELECT path FROM projects WHERE id = ?")
    .get(id) as { path: string } | undefined;

  if (!row) {
    throw new Error(`Project not found: ${id}`);
  }

  return readOrCreateConfig(row.path);
}

/**
 * Write `.aichemist/config.json` for a project.
 */
export function saveProjectConfig(
  db: Database,
  id: string,
  config: ProjectConfig
): void {
  const row = db
    .prepare("SELECT path FROM projects WHERE id = ?")
    .get(id) as { path: string } | undefined;

  if (!row) {
    throw new Error(`Project not found: ${id}`);
  }

  const cfgPath = configPath(row.path);
  const dir = path.dirname(cfgPath);
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(cfgPath, JSON.stringify(config, null, 2), "utf-8");
}

File: src-tauri/src/projects.rs
use rusqlite::params;
use serde::{Deserialize, Serialize};
use std::path::Path;
use uuid::Uuid;

use crate::AppState;

// ─── Data types ──────────────────────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Project {
    pub id: String,
    pub name: String,
    pub path: String,
    pub created_at: String,
    pub config: ProjectConfig,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ProjectConfig {
    pub provider: String,
    pub model: String,
    pub approval_mode: String,
    pub approval_rules: Vec<ApprovalRule>,
    pub custom_tools: Vec<serde_json::Value>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ApprovalRule {
    pub tool_category: String,
    pub policy: String,
}

impl Default for ProjectConfig {
    fn default() -> Self {
        Self {
            provider: "anthropic".into(),
            model: "claude-sonnet-4-5".into(),
            approval_mode: "custom".into(),
            approval_rules: vec![
                ApprovalRule { tool_category: "filesystem".into(), policy: "risky_only".into() },
                ApprovalRule { tool_category: "shell".into(), policy: "always".into() },
                ApprovalRule { tool_category: "web".into(), policy: "never".into() },
            ],
            custom_tools: vec![],
        }
    }
}

// ─── Helpers ─────────────────────────────────────────────────────────────────

fn config_path(project_path: &str) -> std::path::PathBuf {
    Path::new(project_path).join(".aichemist").join("config.json")
}

fn read_or_create_config(project_path: &str) -> ProjectConfig {
    let path = config_path(project_path);
    if path.exists() {
        let raw = std::fs::read_to_string(&path).unwrap_or_default();
        serde_json::from_str(&raw).unwrap_or_default()
    } else {
        let config = ProjectConfig::default();
        if let Some(parent) = path.parent() {
            let _ = std::fs::create_dir_all(parent);
        }
        let _ = std::fs::write(&path, serde_json::to_string_pretty(&config).unwrap());
        config
    }
}

fn now_iso() -> String {
    chrono::Utc::now().to_rfc3339()
}

// ─── Tauri commands ───────────────────────────────────────────────────────────

/// Open a folder as a project. Creates `.aichemist/config.json` if absent.
#[tauri::command]
pub fn add_project(path: String, state: tauri::State<AppState>) -> Result<Project, String> {
    let path = path.trim_end_matches('/').to_string();

    if !Path::new(&path).is_dir() {
        return Err(format!("Path is not a directory: {path}"));
    }

    let name = Path::new(&path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("Project")
        .to_string();

    let config = read_or_create_config(&path);
    let id = Uuid::new_v4().to_string();
    let created_at = now_iso();

    let conn = state.db.lock().map_err(|e| e.to_string())?;

    // INSERT OR IGNORE so re-opening an existing project is a no-op on the DB level
    conn.execute(
        "INSERT OR IGNORE INTO projects (id, name, path, created_at) VALUES (?1, ?2, ?3, ?4)",
        params![id, name, path, created_at],
    )
    .map_err(|e| e.to_string())?;

    // Fetch the canonical row (handles the case where the project already existed)
    let row = conn
        .query_row(
            "SELECT id, name, path, created_at FROM projects WHERE path = ?1",
            params![path],
            |row| {
                Ok((
                    row.get::<_, String>(0)?,
                    row.get::<_, String>(1)?,
                    row.get::<_, String>(2)?,
                    row.get::<_, String>(3)?,
                ))
            },
        )
        .map_err(|e| e.to_string())?;

    Ok(Project {
        id: row.0,
        name: row.1,
        path: row.2,
        created_at: row.3,
        config,
    })
}

/// Return all known projects, ordered by creation date.
#[tauri::command]
pub fn list_projects(state: tauri::State<AppState>) -> Result<Vec<Project>, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;

    let mut stmt = conn
        .prepare("SELECT id, name, path, created_at FROM projects ORDER BY created_at ASC")
        .map_err(|e| e.to_string())?;

    let projects = stmt
        .query_map([], |row| {
            Ok((
                row.get::<_, String>(0)?,
                row.get::<_, String>(1)?,
                row.get::<_, String>(2)?,
                row.get::<_, String>(3)?,
            ))
        })
        .map_err(|e| e.to_string())?
        .filter_map(|r| r.ok())
        .map(|(id, name, path, created_at)| {
            let config = read_or_create_config(&path);
            Project { id, name, path, created_at, config }
        })
        .collect();

    Ok(projects)
}

/// Remove a project from the registry. Does NOT delete the folder.
#[tauri::command]
pub fn remove_project(id: String, state: tauri::State<AppState>) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    conn.execute("DELETE FROM projects WHERE id = ?1", params![id])
        .map_err(|e| e.to_string())?;
    Ok(())
}

/// Read `.aichemist/config.json` for a project.
#[tauri::command]
pub fn get_project_config(id: String, state: tauri::State<AppState>) -> Result<ProjectConfig, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    let path: String = conn
        .query_row("SELECT path FROM projects WHERE id = ?1", params![id], |r| r.get(0))
        .map_err(|e| e.to_string())?;
    Ok(read_or_create_config(&path))
}

/// Write `.aichemist/config.json` for a project.
#[tauri::command]
pub fn save_project_config(
    id: String,
    config: ProjectConfig,
    state: tauri::State<AppState>,
) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    let path: String = conn
        .query_row("SELECT path FROM projects WHERE id = ?1", params![id], |r| r.get(0))
        .map_err(|e| e.to_string())?;
    drop(conn); // release lock before file I/O

    let cfg_path = config_path(&path);
    if let Some(parent) = cfg_path.parent() {
        std::fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    let json = serde_json::to_string_pretty(&config).map_err(|e| e.to_string())?;
    std::fs::write(cfg_path, json).map_err(|e| e.to_string())?;
    Ok(())
}
